File: solutions/cpp/0670.cpp
class Solution {
 public:
  int maximumSwap(int num) {
    string s = to_string(num);
    unordered_map<char, int> map;

    for (int i = 0; i < s.length(); ++i) map[s[i]] = i;

    for (int i = 0; i < s.length(); ++i)
      for (char digit = '9'; digit > s[i]; --digit)
        if (map[digit] > i) {
          s[map[digit]] = s[i];
          s[i] = digit;
          return stoi(s);
        }

    return num;
  }
};

File: solutions/java/0670.java
class Solution {
  public int maximumSwap(int num) {
    char[] s = Integer.toString(num).toCharArray();
    int[] map = new int[10];

    for (int i = 0; i < s.length; ++i)
      map[s[i] - '0'] = i;

    for (int i = 0; i < s.length; ++i)
      for (int j = 9; j > s[i] - '0'; --j)
        if (map[j] > i) {
          s[map[j]] = s[i];
          s[i] = (char) ('0' + j);
          return Integer.valueOf(new String(s));
        }

    return num;
  }
}